### DISCUSSION TITLE
where clause for foreach loop

### DISCUSSION THREAD

**Discussion by xtofs** [dotnet/csharplang#1588] [Language Ideas]
To iterate only over the items of a list with a certain property, I find myself writing:
```
foreach(var item in (from x in items where x > a select x)) 
```
(Added parenthesis only for readability)

I suggest to introduce a `where` clause to the foreach statement analog to Linq expression syntax.

```
foreach(var item in items where item > a)
```

**Comment by Joe4evr** [comment]
What's wrong with this?
```cs
foreach (var item in items.Where(i => i > a))
```

**Comment by bondsbw** [comment]
@Joe4evr Not as clean, since your suggestion has two names for the same variable (`item` and `i`).

**Comment by alrz** [comment]
there's is a proposal to allow linq as a statement
```cs
from item in list
where f(item)
do {
 WriteLine(item);
}
```
#101

**Comment by DavidArno** [comment]
Whilst #101 is good, I think the idea implied by this issue - allow linq query syntax to be used in `foreach` loops - makes for a better solution:
```cs
foreach (var student in students
         where student.Year > 1
         let gpa = gradesDb.Lookup(student).Gpa
         where gpa >= 3.0)
{
    Console.WriteLine($"{student.Name} ${gpa}");
}
```

**Comment by alrz** [comment]
#101 is actually concluded from what you're suggesting (https://github.com/dotnet/roslyn/issues/1938).

**Comment by DavidArno** [comment]
@alrz,

That's a shame. It's local functions all over again from my perspective. Someone suggests a really good idea (implied typed lambdas and linq query syntax in foreach loops), then someone, for non obvious reasons, suggests an inferior solution (local functions and `from .. in .. do`, respectively) and that gathers popular support and ends up being the thing that gets implemented.

Oh well, there's no point in fighting the current...

**Comment by alrz** [comment]
>That's a shame. It's local functions all over again from my perspective. Someone suggests a really good idea (implied typed lambdas and linq query syntax in foreach loops), then someone, for non obvious reasons, suggests an inferior solution (local functions and `from .. in .. do`, respectively) and that gathers popular support and ends up being the thing that gets implemented.

I didn't see how that's related to local functions.


But I agree, it sucks that people have different opinions.

**Comment by DavidArno** [comment]
> But I agree, it sucks that people have different opinions.

I've  no problem with folk having different opinions. It only sucks when the inferior (from my perspective) opinion wins out 😆

**Comment by bondsbw** [comment]
@DavidArno I don't lean strongly either direction.  But consider if a continuation (`into` clause) is added.  Then your `var student` is out of scope in the block, which just has an odd feel.

**Comment by DavidArno** [comment]
@bondsbw, sorry but I've no idea what an `into` clause is. Would you mind showing an example?

**Comment by bondsbw** [comment]
Extending your example:

```c#
foreach (var student in students
         where student.Year > 1
         let gpa = gradesDb.Lookup(student).Gpa
         where gpa >= 3.0
         group student by student.Major into studentsByMajor)
{
    Console.WriteLine($"{student.Name} ${gpa}"); // Errors: student undefined, gpa undefined
}
```
